core/src/connection_reuse.rs

```diff
@@ -40,205 +40,556 @@
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture> + Clone,
+    C::Output: StreamMuxer + 'static,
+    UpgradedNode<T, C>: Transport<Output = C::Output> + Clone,
+    <UpgradedNode<T, C> as Transport>::Dial: 'static,
+    <UpgradedNode<T, C> as Transport>::MultiaddrFuture: 'static,
+{
+    /// Informs the Manager about a new inbound connection that has been
+    /// established, so it
+    fn new_inbound(&self, addr: &Multiaddr, listener_id: usize, muxer: C::Output) {
+        let mut conns = self.connections.lock();
+        let new_state = PeerState::Active {
+            muxer: Arc::new(muxer),
+            listener_id: Some(listener_id),
+            closed: Arc::new(AtomicBool::new(false))
+        };
+        let mut old_state = conns.insert(addr.clone(), new_state);
+
+        if let Some(PeerState::Pending { ref mut notify, .. }) = old_state {
+            // we need to wake some up, that we have a connection now
+            trace!("Found incoming for pending connection to {}", addr);
+            for (_, task) in notify.drain() {
+                task.notify()
+            }
+        } else {
+            old_state.map(|s| s.reset_conn());
+        }
+    }
+
+    /// Polls for the outbound stream of `addr`. Clears the cached value first if `reset` is true.
+    /// Dials, if no connection is in the internal cache. Returns `Ok(Async::NotReady)` as long as
+    /// the connection isn't establed and ready yet.
+    fn poll_outbound(
+        &self,
+        addr: &Multiaddr,
+    ) -> Poll<(SubstreamRef<Arc<C::Output>>, Arc<AtomicBool>), IoError> {
+        let mut conns = self.connections.lock();
+
+        let state = conns.entry(addr.clone()).or_insert_with(|| {
+            let state = match self.transport.clone().dial(addr.clone()) {
+                Ok(future) => {
+                    trace!("Opening new connection to {:?}", addr);
+                    let future = future.and_then(|(out, addr)| addr.map(move |a| (out, a)));
+                    let future = Box::new(future);
+
+                    let mut notify = HashMap::<usize, task::Task, UsizeHasherBuilder>::default();
+                    // make sure we are woken up once this connects
+                    notify.insert(TASK_ID.with(|&t| t), task::current());
+
+                    PeerState::Pending { future, notify }
+                }
+                Err(_) => {
+                    trace!(
+                        "Failed to open connection to {:?}, multiaddr not supported",
+                        addr
+                    );
+                    let err =
+                        IoError::new(IoErrorKind::ConnectionRefused, "multiaddr not supported");
+                    PeerState::Errored(err)
+                }
+            };
+            self.new_con_notify();
+            // Althought we've just started in pending state, there are transports, which are
+            // immediately ready - namely 'memory' - so we need to poll our future right away
```

Comment:
You always need to poll immediately, otherwise the current task will not be waken up later when it needs to.